crates/uv-pep440/src/version_ranges_specifier.rs

```diff
@@ -1,66 +1,81 @@
+use std::error::Error;
+use std::fmt::{Display, Formatter};
 use std::ops::Bound;
 
-use itertools::Itertools;
-use thiserror::Error;
 use version_ranges::Ranges;
 
-use uv_pep440::{Operator, Prerelease, Version, VersionSpecifier, VersionSpecifiers};
+use crate::{Operator, Prerelease, Version, VersionSpecifier, VersionSpecifiers};
 
-#[derive(Debug, Error)]
-pub enum PubGrubSpecifierError {
-    #[error("~= operator requires at least two release segments: `{0}`")]
+/// The conversion between PEP 440 [`VersionSpecifier`] and version-ranges
+/// [`VersionRangesSpecifier`] failed.
+#[derive(Debug)]
+pub enum VersionRangesSpecifierError {
+    /// The `~=` operator requires at least two release segments
     InvalidTildeEquals(VersionSpecifier),
 }
 
+impl Error for VersionRangesSpecifierError {}
+
+impl Display for VersionRangesSpecifierError {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        match self {
+            Self::InvalidTildeEquals(specifier) => {
+                write!(
+                    f,
+                    "The `~=` operator requires at least two release segments: `{specifier}`"
```

Comment:
@BurntSushi Can this error ever happen / could we catch that in the parser and make the conversion infallible?

Reply:
You mean the parser for marker expressions? I think there are tests covering it?

https://github.com/astral-sh/uv/blob/36102dbd0ef08b5bc9b8ab251aa57e9d645584d3/crates/uv-pep508/src/marker/tree.rs#L2053

Whether `~=` is actually used in the real world, I'm not sure. I don't recall seeing it.

(I'm not 100% certain I've understood everything correctly here.)

Oh, I think I did misunderstand. Perhaps it's the _combination_ of `~=` and a lack of two release segments. Indeed, PEP 440 says:

> This operator MUST NOT be used with a single segment version number such as ~=1.

So I guess we could indeed catch this at parse time.

A secondary problem is how to represent this in the type system. Otherwise you'll need a panicking branch here instead of an error. Which I think seems fine as long as we document our guarantees.